### PR TITLE
PYPI proof-of-concept

### PR DESCRIPTION
**Pull request by novicecpp**
Proof-of-concept of https://github.com/dmwm/CRABServer/issues/7560

Please see design choices https://github.com/dmwm/CRABServer/issues/7560#issuecomment-1859531116 and walkthrough of both images https://github.com/dmwm/CRABServer/issues/7560#issuecomment-1916881209 https://github.com/dmwm/CRABServer/issues/7560#issuecomment-1923506222

### CODE REVIEW

**Comment by novicecpp** [comment]
@belforte let's go back to the issue in https://github.com/dmwm/CRABServer/issues/7560 and discuss there.
Will follow up later, after new year.

File: cicd/crabserver_pypi/requirementsParse.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+# wmcore @ git+https://github.com/dmwm/WMCore@2.2.4rc6
+parser = argparse.ArgumentParser()
+parser.add_argument('-f', '--filename', default='wmcore_requirements.txt')
+args = parser.parse_args()
+with open(args.filename, encoding='utf-8') as r:
+    wmcoreRequirementStr = r.readlines()[0]
+_, tmpRepo, tmpCommit = wmcoreRequirementStr.split('@')
+commit = tmpCommit.strip()
+repo = tmpRepo.split('+')[1].strip()
+print(repo, commit)
```
Comment (mapellidario) [inline cicd/crabserver_pypi/requirementsParse.py:13]: could you add an example of the desired output? thanks!

File: cicd/crabserver_pypi/manage [hunk 2]
```diff
@@ -0,0 +1,107 @@
+#!/bin/bash
+##H Usage: manage ACTION [ATTRIBUTE] [SECURITY-STRING]
+##H
```
Comment (mapellidario) [inline cicd/crabserver_pypi/manage:3]: for example, here I would add a pointer tohttps://github.com/dmwm/deployment/blob/master/crabserver/manage

File: cicd/crabtaskworker_pypi/runContainer.sh.patch [hunk 3]
```diff
@@ -0,0 +1,16 @@
+--- runContainer_pypi.sh        2023-12-04 22:49:56.106418559 +0100
```
Comment (mapellidario) [inline cicd/crabtaskworker_pypi/runContainer.sh.patch:1]: maybe  this is not necessary?

**Review by mapellidario — CHANGES_REQUESTED** [review]
Thanks Wa, this is terrific work! :)

I have a couple of comments and requests that are mainly intended to help me follow along. I am not really concerned with how you implement the building system, I totally trust your judgement for that. I am more interested about where we install our dependencies from and which part of the building system you are reusing and I would ask you to add some information to help me understand.

Regarding the dependencies, I know it is all in the dockerfiles, but the devil is in the details and I fear I miss it. Could you provide a list/table that summarizes where do we get dependencies from, when they are not from pypi?

for example: crabserver:

- tnsnames. before: they were already installed in the docker image from cmsweb, now: same source, `registry.cern.ch/cmsweb/cmsweb:20230427-stable`

crabtaskworker:

- gfal. before: ??? (i can not quickly find it), now: `registry.cern.ch/cmsweb/gfal:latest`
- ldap config. before ???, now: `gitlab-registry.cern.ch/linuxsupport/cc7-base:20231001-1.x86_64`
- ...

Or maybe, if you have it, you can just add in the current dockerfile a pointer to where the same dependency was taken from in the rpm build system. Or specify that you did not have to take care of it because it was sort of available for "free".

Regarding the build system, could you add a comment with a pointer to the original source in to the files that you are reusing (such as the `manage` script) or that you are adapting (such as `new_htcondor_make_runtime.sh`)? Thanks, sorry it takes a bit of time to switch from the code of your PR and the comment where you summarized here each file is from. I would prefer to have "self-documenting" code :) (and i consider the information "this files originally comes from <here>" to be documentation)

In order to help me reproduce your build, could you add a simple "build.sh" script with the proper `docker build` parameters, and maybe also an indication which directory those commands should be executed from? thanks! Or maybe a pointer to your jenkins / gitlab pipelines should be enough.

Thanks man, sorry if my review is overdue

**Comment by novicecpp** [comment]
(Let me giving some quick reply that I have in my mind right now. I will provide in-detail answer later after I come back)

First of all, thanks for your comments. I really appreciate it.

- About the dependencies, I will fix the documentation.
It will make you more understand on how our application work and what change I introduce in this PR.
- All scripts here should be independent from original copy and self-documents. I can add ref to the original file I copied from in separate docs, what you see now is the result of my bad documenting.
- I still think it is needed to go through Dockerfile and understand each step by yourself. Of course, it is my job to make it easier to read, I will fix that.
- The `build.sh` you want is described in https://github.com/dmwm/CRABServer/issues/7560#issuecomment-1859531116 in *1. Building* section, please have a look. 
    - In summary, pull this PR, cd to git root dir, then simply run `docker build` command without passing any build args.

**Comment by novicecpp** [comment]
I need to test in devthree later.
But it should have no more changes unless a bug has been found.

File: cicd/crabserver_pypi/Dockerfile [hunk 4]
```diff
@@ -25,7 +27,7 @@ WORKDIR /build
 COPY cicd/crabserver_pypi/ .
 # Getting wmcore version from wmcore_requirements.txt and install with `pip install --no-deps`
 RUN pip install -r requirements.txt
-RUN req=($(./requirementsParse.py wmcore_requirements.txt)) \
+RUN req=($(grep -v '^\s*#' < wmcore_requirements.txt)) \
```
Comment (belforte) [inline cicd/crabserver_pypi/Dockerfile:30]: for dummies like me, can you add a one line comment to indicate what it does ? I am puzzled by `\s`

Comment (novicecpp) [reply cicd/crabserver_pypi/Dockerfile:32]: Done.
(sorry, I am force-pushing with new comment lines make GH comment point to new changes instead of old one).

**Comment by novicecpp** [comment]
@belforte @mapellidario PR is ready for review.
 All status tracking tests passed (running in devthree env).
```
240212_105438:tseethon_crab_20240212_115438
240212_105440:tseethon_crab_20240212_115439
240212_105441:tseethon_crab_20240212_115441
240212_105443:tseethon_crab_20240212_115443
240212_105447:tseethon_crab_20240212_115445
240212_105449:tseethon_crab_rucio_transfers_20240212_115448
240212_105451:tseethon_crab_rucio_transfers_group_20240212_115451
240212_105454:tseethon_crab_rucio_transfers_manyedm_nopublication_20240212_115452
240212_105457:tseethon_crab_rucio_transfers_nopublication_20240212_115455
240212_105500:tseethon_crab_rucio_transfers_withnoneedm_20240212_115458
240212_105502:tseethon_crab_rucio_transfers_withnoneedm_nopublication_20240212_115501'
```
For example, [240212_105441:tseethon_crab_20240212_115441](https://cmsweb-test12.cern.ch/crabserver/ui/task/240212_105443%3Atseethon_crab_20240212_115443) ([Grafana](https://monit-grafana.cern.ch/d/cmsTMDetail/cms-task-monitoring-task-view?orgId=11&var-user=tseethon&var-task=240212_105441%3Atseethon_crab_20240212_115441&from=1707731681000&to=now)).

File: cicd/crabserver_pypi/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+#RUN mkdir -p /opt/rucio/etc/
+
+RUN mkdir /build
+WORKDIR /build
+COPY cicd/crabserver_pypi/ .
+# Install wmcore/crab dependencies
+RUN pip install -r requirements.txt
+# Getting the version number from wmcore_requirements.txt and install it with `--no-deps`.
+# When reading from file, `grep -v '^\s*#'` to ignore all comment lines, including the ones with leading whitespace ('\s' to match whitespace).
+RUN req=($(grep -v '^\s*#' < wmcore_requirements.txt)) \
```
Comment (belforte) [inline cicd/crabserver_pypi/Dockerfile:32]: this bash is still too much advanced for me to understand. I can run it and see that it works, but ($(cmd)) it is out of my league. Oh well...
I guess you do not like using `cut` or `awk` to parse the two elements of `https://github.com/dmwm/WMCore 2.2.4rc6`.

File: cicd/crabserver_pypi/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    && pip install --no-deps WMCore==${req[1]}
+
```
Comment (belforte) [inline cicd/crabserver_pypi/Dockerfile:34]: Why do we need to pip-install WMCore once we start form wmagent-base image ? What is missing in there that this adds ?

File: cicd/crabserver_pypi/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+# Install CRAB.
+COPY src/python/ /data/srv/current/lib/python3.8/site-packages
```
Comment (belforte) [inline cicd/crabserver_pypi/Dockerfile:36]: it looks fragile that we have `python3,.8` here but not elsewhere. Maybe add a comment that i needs to match the python in the base image we start from ? Or is there a way to detect at build time which python version we are using ? I also notice "hardcoded"| `python3.8` in `cicd/crabserver_pypi/manage` at L41. Maybe the `3.8` is simply not needed ?

**Comment by belforte** [comment]
sorry, I mixed single comments and review comments... sloppy clicking :-(

**Comment by belforte** [comment]
Thanks @novicecpp . I am done looking at this. I can not claim I really "reviewed" but since 1. it works 2. is new and will not break anything, I did not look for mistakes, only tried to have something that I can read an understand at least in broad terms.

File: cicd/crabserver_pypi/Dockerfile (hunk 5 above)
Comment (novicecpp) [reply cicd/crabserver_pypi/Dockerfile:32]: It is the bash built-in on how you create array (put string separate by space inside parenthesis and assign to variable).

`cut` is also fine for me. What about 
```bash
wmcore_version=$(grep -v '^\s*#' wmcore_requirements.txt | cut -d' ' -f2)
```
?

File: cicd/crabserver_pypi/Dockerfile (hunk 7 above)
Comment (novicecpp) [reply cicd/crabserver_pypi/Dockerfile:36]: The python version path is managed by `pip`, and in runtime python look at the path that match its version. 
In PoC, I do that to mock what the path would look like when we convert to pip install later.
But I think simplify to `/data/srv/current/site-packages` is much better, we know `site-packages` is python code, and no need python version because we do not have multiple python version in single container.

**Comment by novicecpp** [comment]
@belforte @mapellidario may I merge this?

**Review by mapellidario — APPROVED** [review]
Thanks wa for the terrific work, you can merge :)

**Review by belforte — APPROVED** [review]

**Comment by belforte** [comment]
sorry for delayed action. Something new always happens here. Please go ahead

File: cicd/crabtaskworker_pypi/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+FROM python:3.8 as build-data
+SHELL ["/bin/bash", "-c"]
+RUN mkdir /build \
+    && apt-get update \
+    && apt-get install -y curl zip git \
+    && apt-get clean all
+WORKDIR /build
+COPY cicd/crabtaskworker_pypi/new_htcondor_make_runtime.sh /build
+COPY . /build
+COPY --from=wmcore-src /WMCore /build/WMCore
```
[the remark this answers is not in the document]
Comment (novicecpp) [reply cicd/crabtaskworker_pypi/Dockerfile:22]: It is like you have 3 directory:
- `wmcore-src` (Line 2): you copy (e.g. `cp/rsync`) everything from `python:3.8` directory, created by some guy on the internet, to your `wmcore-src`, then put your own files in there
- `build-data` (Line 12): same procedure, you copy everything from `python:3.8` to your `build-data`, this time you put another set of files in it, plus copy files from `wmcore-src` directory to this directory too.
- main directory (last FROM statement, without name, Line 42): you copy everything from  `wmagent-base:pypi-20230705`, you put yet-another-set-of-files in there, plus files from `build-data`
    - We do not care the contents in `wmcore-src`.

At the end of `docker build`, the first 2 directory get discarded. The main directory, you tag it with `crabtaskworker:manual`, zip it, and upload to docker hub.

File: cicd/crabtaskworker_pypi/runContainer.sh.patch
[the remark this answers is not in the document]
Comment (novicecpp) [reply cicd/crabtaskworker_pypi/runContainer.sh.patch:1]: See comment in https://github.com/dmwm/CRABServer/pull/8088/files/46ff962b1d6de17db7cca8cc08425ad602e26cf5#diff-6800f12d5d596f7ac80b23d4e7f5440cdf8db3e853bf209c154ff3547a758f66

**Comment by novicecpp** [comment]
Thanks. Merging...
